numpy/core/src/_simd/_simd.c

```diff
@@ -0,0 +1,43 @@
+#include "_simd.h"
+
+PyMODINIT_FUNC PyInit__simd(void)
+{
+    static struct PyModuleDef defs = {
+        .m_base = PyModuleDef_HEAD_INIT,
+        .m_name = "_simd",
+        .m_size = -1
+    };
+    if (npy_cpu_init() < 0) {
+        return NULL;
+    }
+    PyObject *m = PyModule_Create(&defs);
+    if (m == NULL) {
+        goto err;
+    }
+    PyObject *targets = PyDict_New();
+    if (targets == NULL) {
+        goto err;
+    }
+    NPY_CPU_DISPATCH_CALL_ALL(simd_attach_module, (targets))
+    // Add keys for non-supported optimizations with None value
+    #define CHECK_NON_SUPPORTED(TESTED_FEATURES, TARGET_NAME, ...) \
+        if (!TESTED_FEATURES) { \
+            Py_INCREF(Py_None); \
+            if (PyDict_SetItemString(targets, NPY_TOSTRING(TARGET_NAME), Py_None) < 0) { \
+                Py_DECREF(Py_None); Py_DECREF(targets); goto err; \
+            } \
+        }
+    NPY__CPU_DISPATCH_CALL(NPY_CPU_HAVE, CHECK_NON_SUPPORTED, DUMMY_4MSVC)
+
+    if (PyErr_Occurred() || PyModule_AddObject(m, "targets", targets) < 0) {
+        Py_DECREF(targets);
+        goto err;
+    }
+    return m;
+err:
+    if (!PyErr_Occurred()) {
+        PyErr_SetString(PyExc_RuntimeError, "cannot load _simd module.");
+    }
```

Comment:
```suggestion
    if (PyErr_Occurred()) {
        // can this happen inside NPY__CPU_DISPATCH_CALL?
        PyErr_SetString(PyExc_RuntimeError, "cannot load _simd module.");
        Py_DECREF(targets);
        goto err;
    }
    if (PyModule_AddObject(m, "targets", targets) < 0) {
        Py_DECREF(targets);
        goto err;
    }
    return m;
err:
```

Reply:
not  '!' operator is missing
```Diff
--if (PyErr_Occurred()) {
++if (!PyErr_Occurred()) {
````
All calls that reached by `NPY__CPU_DISPATCH_CALL` are exposed within the preprocessor callback `CHECK_NON_SUPPORTED` and all of them are pyapi calls.

> // can this happen inside NPY__CPU_DISPATCH_CALL?

No, I reviewed the pyapi calls within `CHECK_NON_SUPPORTED` and I think nothing wrong on it

> If you separate these conditions you no longer need that branch in the err path

The existence behind this check is to detect any unexpected behaviors that including any pyapi calls,
so I still prefer to keep it as-is.

Reply:
> The existence behind this check is to detect any unexpected behaviors that including any pyapi calls,
so I still prefer to keep it as-is.

Python already handles this for us and raises `SystemError` if we screw up. We should not be writing code defensively against the python API not doing what it says it does, that just makes it harder to reason what is actually happening.

Reply:
> No, I reviewed the pyapi calls within CHECK_NON_SUPPORTED and I think nothing wrong on it

Right, in that case the `PyErr_Occurred()` occurred call should be just after calling `simd_attach_module`. There are only three cases you ever need to call `PyErr_Occurred()`:

* you actually want information about the exception itself
* you are calling a weird python API that specifically mentions the need to use `PyErr_Occurred` (i.e., distinguishing the int `-1` from an error converting an integer)
* you are calling one of our own function that is unable to return an error easily (eg ufunc inner loops, `simd_attach_module` etc

Note that "just in case" is not on this list :)

Reply:
> Note that "just in case" is not on this list :)

I concluded it from NumPy multiarray module code, but I like your list and it makes sense.
Well, in that case, all I have to do is removing:
```Diff
err:
-    if (!PyErr_Occurred()) {
-        PyErr_SetString(PyExc_RuntimeError, "cannot load _simd module.");
-    }
````

It seems I have to stop build conclusions